flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/AsyncIntervalJoinOperator.java

```diff
@@ -0,0 +1,428 @@
+                        () -> Long.MIN_VALUE);
+
+        collector = new TimestampedCollectorWithDeclaredVariable<>(output, resultTimestamp);
+        context = new ContextImpl(userFunction, resultTimestamp, leftTimestamp, rightTimestamp);
+        internalTimerService =
+                getInternalTimerService(CLEANUP_TIMER_NAME, StringSerializer.INSTANCE, this);
+    }
+
+    /**
+     * Process a {@link StreamRecord} from the left stream. Whenever an {@link StreamRecord} arrives
+     * at the left stream, it will get added to the left buffer. Possible join candidates for that
+     * element will be looked up from the right buffer and if the pair lies within the user defined
+     * boundaries, it gets passed to the {@link ProcessJoinFunction}.
+     *
+     * @param record An incoming record to be joined
+     * @throws Exception Can throw an Exception during state access
+     */
+    @Override
+    public void processElement1(StreamRecord<T1> record) throws Exception {
+        processElement(record, leftBuffer, rightBuffer, lowerBound, upperBound, true);
+    }
+
+    /**
+     * Process a {@link StreamRecord} from the right stream. Whenever a {@link StreamRecord} arrives
+     * at the right stream, it will get added to the right buffer. Possible join candidates for that
+     * element will be looked up from the left buffer and if the pair lies within the user defined
+     * boundaries, it gets passed to the {@link ProcessJoinFunction}.
+     *
+     * @param record An incoming record to be joined
+     * @throws Exception Can throw an exception during state access
+     */
+    @Override
+    public void processElement2(StreamRecord<T2> record) throws Exception {
+        processElement(record, rightBuffer, leftBuffer, -upperBound, -lowerBound, false);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <THIS, OTHER> void processElement(
+            final StreamRecord<THIS> record,
+            final MapState<Long, List<IntervalJoinOperator.BufferEntry<THIS>>> ourBuffer,
+            final MapState<Long, List<IntervalJoinOperator.BufferEntry<OTHER>>> otherBuffer,
+            final long relativeLowerBound,
+            final long relativeUpperBound,
+            final boolean isLeft)
+            throws Exception {
+
+        final THIS ourValue = record.getValue();
+        final long ourTimestamp = record.getTimestamp();
+
+        if (ourTimestamp == Long.MIN_VALUE) {
+            throw new FlinkException(
+                    "Long.MIN_VALUE timestamp: Elements used in "
+                            + "interval stream joins need to have timestamps meaningful timestamps.");
+        }
+
+        if (isLate(ourTimestamp)) {
+            sideOutput(ourValue, ourTimestamp, isLeft);
+            return;
+        }
+
+        addToBuffer(ourBuffer, ourValue, ourTimestamp)
+                .thenCompose((empty) -> otherBuffer.asyncEntries())
+                .thenAccept(
```

Comment:
`thenAccept` -> `thenCompose`?
Because we should wait for the State Iterator to complete.

Reply:
Good catch. Modified accordingly